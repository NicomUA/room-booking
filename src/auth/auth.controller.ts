import { Body, Controller, Post } from '@nestjs/common';
import { AuthService } from './auth.service';
import { LoginDto } from './dto/login.dto';
import { JwtPayloadResponse } from './dto/payload.dto';

@Controller()
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('login')
  async login(
    @Body() { email, password }: LoginDto,
  ): Promise<JwtPayloadResponse> {
    return this.authService.login({ email, password });
  }
}
