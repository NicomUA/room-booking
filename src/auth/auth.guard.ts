import { ConfigurationService } from '@app/configuration';
import {
  CanActivate,
  ExecutionContext,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { Request } from 'express';
import { JWTPayload } from './dto/payload.dto';

@Injectable()
export class AuthGuard implements CanActivate {
  constructor(
    private jwtService: JwtService,
    private config: ConfigurationService,
  ) {}

  async canActivate(context: ExecutionContext): Promise<boolean> {
    const request = context.switchToHttp().getRequest();
    const token = this.extractTokenFromHeader(request);
    const JWT_SECRET = this.config.get('JWT_SECRET');

    if (!token) {
      throw new UnauthorizedException('Token is missing');
    }
    try {
      const payload = await this.jwtService.verifyAsync<JWTPayload>(token, {
        secret: JWT_SECRET,
      });
      request['user'] = { id: payload.sub, ...payload };
    } catch {
      throw new UnauthorizedException();
    }
    return true;
  }

  private extractTokenFromHeader(request: Request): string | undefined {
    const [type, token] = request.headers.authorization?.split(' ') ?? [];
    return type === 'Bearer' ? token : undefined;
  }
}
