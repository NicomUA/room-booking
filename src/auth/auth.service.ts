import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { LoginDto } from './dto/login.dto';
import { JwtPayloadResponse } from './dto/payload.dto';
import { UserService } from '../user/user.service';

@Injectable()
export class AuthService {
  constructor(
    private jwtService: JwtService,
    private userService: UserService,
  ) {}

  async login({ email, password }: LoginDto): Promise<JwtPayloadResponse> {
    const user = await this.userService.findByEmail(email);

    if (!user) {
      throw new NotFoundException('User not found');
    }

    if (user.password !== password) {
      throw new BadRequestException('Wrong password');
    }

    const payload = { sub: user.id, email: user.email, name: user.name };
    return {
      access_token: await this.jwtService.signAsync(payload),
    };
  }
}
