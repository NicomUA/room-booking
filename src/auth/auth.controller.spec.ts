import { AuthController } from './auth.controller';
import { AuthService } from './auth.service';
import { TestBed } from '@automock/jest';

describe('AuthController', () => {
  let controller: AuthController;
  let authService: jest.Mocked<AuthService>;

  beforeAll(async () => {
    const { unit, unitRef } = TestBed.create(AuthController).compile();
    controller = unit;
    authService = unitRef.get<AuthService>(AuthService);
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  describe('login', () => {
    it('should login', async () => {
      authService.login.mockResolvedValue({ access_token: 'test' });

      await expect(
        await controller.login({ email: 'test', password: 'test' }),
      ).toEqual({
        access_token: 'test',
      });
    });

    it('should throw an error', async () => {
      authService.login.mockImplementation(() => {
        throw new Error();
      });

      await expect(
        controller.login({ email: 'test', password: 'test' }),
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
