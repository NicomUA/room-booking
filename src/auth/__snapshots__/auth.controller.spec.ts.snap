// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthController login should throw an error 1`] = `""`;
